lib/plugins/backup.rb

```diff
@@ -0,0 +1,124 @@
+require 'rubygems/package'
+require 'zlib'
+require_relative '../moonshot/creds_helper'
+
+module Moonshot::Plugins # rubocop:disable Documentation
+  # Monkeypatching TarWriter for our
+  # needs to improve clarity
+  refine Gem::Package::TarWriter do
+    def add(path:, name:, permission: nil)
+      rel_path = File.join(path << name)
+      permission ||= 0644
+
+      add_file(name, permission) do |io|
+        File.open(rel_path, 'r') { |f| io.write(f.read) }
+      end
+    end
+  end
+end
+
+using Moonshot::Plugins
+
+module Moonshot
+  module Plugins
+    # Moonshot plugin class for deflating and
+    # uploading files on given hooks
+    class Backup
+      include Moonshot::CredsHelper
+
+      attr_accessor :bucket,
+                    :files,
+                    :hooks,
+                    :target_name
+
+      def initialize
+        yield self if block_given?
+        raise ArgumentError if @bucket.nil? || @files.nil? || @hooks.nil?
+      end
+
+      def backup(resources)
+        @app_name = resources.stack.app_name
+        @stack_name = resources.stack.name
+        @target_name ||= '%{app_name}_%{timestamp}_%{user}.tar.gz'
+
+        tar_out = tar @files
+        zip_out = zip tar_out
+        upload zip_out
+      rescue StandardError => e
+        raise e
+      ensure
+        tar_out.close unless tar_out.nil?
+        zip_out.close unless zip_out.nil?
+      end
+
+      # dynamically responding to hooks
+      # supplied in the constructor
+      def method_missing(method_name, *args, &block)
+        backup(*args) if @hooks.include?(method_name) || super
```

Comment:
I think the precedence here might not be what you expect. For example, if `@hooks.include?(method_name)` is false, super will be called. If super returns truthy, `backup(*args)` will be called. I think what you want here is more like:
```ruby
if @hooks.include?(method_name)
  backup(*args)
else
  super
end
```

As an example:
```ruby
def do_thing
	puts "I did a thing!"
end

def have_thing
	false
end

def other_thing
	puts "I did another thing!"
	true
end

do_thing if have_thing || other_thing
```

This code produces:
```
I did another thing!
I did a thing!
=> nil
```

I think you expected it to produce only `I did another thing!`? Of course if the super returned falsey, that would be all that was produced.

In my experience, it's best to be very clear and verbose with code. Optimize for humans :-).